Enhance unified endpoint discovery proposal (#4070)
thanos-io/thanos: docs/proposals/202101_endpoint_discovery.md

title: Unified Endpoint Discovery
type: proposal
menu: proposals
status: accepted
owner: lilic
---

### Related Tickets

* https://github.com/thanos-io/thanos/issues/2600
* https://github.com/thanos-io/thanos/issues/3538

### Summary

We want to propose a new flag called `--endpoint=<address>` that will be passed to Thanos query and replace the current `--store=<address>` and `--rule=<address>` flags, and add a discovery mechanism which will be used to determine what types of services this endpoint can serve.

### Motivation

Currently, in Thanos Query, the discovery of rules APIs happens via Store API's Info method. This makes it harder if we ever want to not have a coupling to the Store API (which is already planned for [scalable ruler proposal](https://github.com/thanos-io/thanos/blob/main/docs/proposals/202005_scalable-rule-storage.md)).

> We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>`. If users use both flags, the Query component performs DNS discovery on often the same address multiple times, this can cause occasional DNS issues as this resolving happens so frequently. This is especially confusing when one works and the other doesn't. Adding new APIs in the future would exacerbate this issue with DNS requests.

Reviewer: ```suggestion
We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>`. If users use both flags, the Query component performs DNS discovery on often the same address multiple times, which can cause occasional DNS issues as this resolution happens so frequently. This is especially confusing when one DNS lookup works and the other doesn't. Adding new APIs in the future would exacerbate this issue with DNS requests.
```

Currently, in Thanos Query, the discovery of rules APIs happens via Store API's Info method. This makes it harder if we ever want to not have a coupling to the Store API (which is already planned for [scalable ruler proposal](https://github.com/thanos-io/thanos/blob/master/docs/proposals/202005_scalable-rule-storage.md)).

We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>`. If users use both flags, the Query component performs DNS discovery on often the same address multiple times, this can cause occasional DNS issues as this resolving happens so frequently. This is especially confusing when one works and the other doesn't. Adding new APIs in the future would exacerbate this issue with DNS requests.

### Goals

Unite discovery of endpoints, which would make code cleaner as well as provide building blocks to provide discovery and support for new gRPC services, for example, the Targets API/Exemplar API/Metadata API.

### Solution

Add a new flag called `--endpoint` to Thanos query, and auto-discover what services that endpoint is serving based on metadata each gRPC server exposes.

> Each component will expose an Info service, that includes various metadata listed below. Discovery of endpoints will happen via this Info service. there might be a case that the discovery of gRPC will also have to happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md), this will be clear once implementation starts.

Reviewer 2: ```suggestion
Each component will expose an Info service, that includes various metadata listed below. Discovery of endpoints will happen via this Info service. There might be a case that the discovery of gRPC will also have to happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md), this will be clear once implementation starts.
```

> Each component will expose an Info service, that includes various metadata listed below. Discovery of endpoints will happen via this Info service. there might be a case that the discovery of gRPC will also have to happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md), this will be clear once implementation starts.

Reviewer 3: > Discovery of endpoints will happen via this Info service.

I think we agreed on using gRPC reflection to discover what services a particular endpoint is exposing? I yes then maybe we should modify the statement to something like 
```suggestion
Each component will expose an Info service, that includes various metadata listed below. Discovery of what services an endpoint is exposing will happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md). Once we discover what services that endpoint is exposing (e.g. StoreAPI, TargetsAPI) we can retrieve metadata about them using a single `Info` method call from the Info service.
```

Author: Yes, I missed it. Thank you for pointing out :)

We also require passing two different flags to the Thanos Query component `--store=<address>` and `--rule=<address>`. If users use both flags, the Query component performs DNS discovery on often the same address multiple times, this can cause occasional DNS issues as this resolving happens so frequently. This is especially confusing when one works and the other doesn't.

> Adding new APIs in the future alike.

Reviewer 2: This sound a bit weird. `alike` adverb means:

```
in the same or a similar way.
```

So it is missing extra information. Perhaps:

```suggestion
Adding new APIs in the future would exacerbate this issue.
```

Author: I think we should add this in the previous paragraph because it is also talking about the same thing?

### Goals

Unite discovery of endpoints, which would make code cleaner as well as provide building blocks to provide discovery and support for new gRPC services, for example, the Targets API/Exemplar API/Metadata API.

### Solution

Add a new flag called `--endpoint` to Thanos query, and auto-discover what services that endpoint is serving based on metadata each gRPC server exposes.

> Each component will expose an Info service, that includes various metadata listed below. Discovery of endpoints will happen via this Info service, there might be a case that the discovery of gRPC will also have to happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md), this will be shown when implementation starts.

Reviewer 2: ```suggestion
Each component will expose an Info service, that includes various metadata listed below. Discovery of endpoints will happen via this Info service, there might be a case that the discovery of gRPC will also have to happen via [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md), this will be clear once implementation starts.
```

    blah: ...
```

> Right now this info service would be added to existing components Ruler, Store, Sidecar, Compact, Downsample, Receive.

Reviewer 2: `bucket downsample` doesn't expose anything over gRPC now (and probably will never do that) so I think we shouldn't mention it here: 
```suggestion
Right now this Info service would be added to existing components Ruler, Store, Sidecar, Compact, Receive.
```

Reviewer 4: I have the same question. How do we expose gRPC on compact? For what APIs?

Reviewer 3: Yeah, I think this was just a slight oversight on our part, there's no plan to have a gRPC API on Compactor yet.

Author: yes agree with @onprem, It will not affect any component not exposing anything over gRPC. we will remove both compact and downsample.

Reviewer 3: And we are missing Querier here too, as it also exposes a Store API we should add Info service there as well. Maybe not mentioning the exact component names will be better.
```suggestion
Right now this Info service would be added to existing components exposing a gRPC server.
```

> The upgrade path would be possible in this case, as it's strictly additive for endpoints, and there will be a migration path to fall back to current discovery methods which could be removed in the future.

Reviewer: ```suggestion
The upgrade path would be possible in this case, as it's strictly additive for endpoints, and there will be a migration path to fall back to current discovery methods, which could be removed in the future.
```

### Alternatives

Initial discovery of gRPC services available will happen via gRPC reflection. To get further information, we would be adding an Info method to other services for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.

This alternative solution was also considered, however it was not chosen because of its following drawbacks.

> 1. For servers implementing multiple API (e.g. Store, Rules, Exemplar), we need to make multiple microservices calls (For given example 3x calls).

Reviewer: ```suggestion
1. For servers implementing multiple APIs (e.g. Store, Rules, Exemplar), we need to make multiple gRPC calls, e.g. three calls for a component implement Store, Rules and Exemplars APIs.
```

> 1. For servers implementing multiple API (Eg. Store, Rules, Exemplar), we need to make multiple microservices calls (For given example 3x calls).

Reviewer 2: ```suggestion
1. For servers implementing multiple API (e.g. Store, Rules, Exemplar), we need to make multiple microservices calls (For given example 3x calls).
```

2. There is also ambiguity about which Info method should be used for these kinds of servers.

### Upgrade plan

> Regardless of the solution, we will be not removing any flags for a given period and have a fallback code in place to be able to discover in the same way as we do right now. This will ensure a smooth upgrade path for the users that make use of the current flags for discovery. We will mark the current flags as deprecated and after two releases have passed we can remove the existing store and rule flag from the query component and remove any migration code.

Reviewer: ```suggestion
Regardless of the solution, we will be not removing any flags for a given period and have a fallback in place to be able to discover in the same way as we do right now. This will ensure a smooth upgrade path for the users who make use of the current flags for discovery. We will mark the current flags as deprecated and after two releases have passed we can remove the existing store and rule flag from the query component and remove any migration code.
```

### Work plan

1. We would be adding the new Info service and flags needed for the above solution, so this part would be entirely additive.

> 2. Second step would be the integration of this service, flags and migration of existing code.

Reviewer: ```suggestion
2. Second step would be the integration of this service, the addition of new flags flags, and migration of existing code.
```

    blah: ...
```

Right now this info service would be added to existing components Ruler, Store, Sidecar, Compact, Downsample, Receive.

The upgrade path would be possible in this case, as it's strictly additive for endpoints, and there will be a migration path to fall back to current discovery methods which could be removed in the future.

### Alternatives

> The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to other services for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.

Reviewer 2: ```suggestion
The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to other services, for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.
```

Reviewer 2: So, will it happen via the Info service or gRPC reflection? I'm confused about this sentence :smile: Maybe I am missing/misunderstanding something. I thought with this proposal we will only use the Info service and we are going to try avoiding reflection

Author: Yes. We are only planning to use the Info service.
During discussions, It was also proposed that we can add Info methods (similar to the Info method we currently have in store API) to all other services (like rules, targets...) instead of creating a new Info service. Discovery of these methods may be done using gRPC reflection ( still depends on implementation 😅 ). That's why I added this as an alternative solution. 

I am also not sure about should we mention this solution as alternative solution? WDYT :)

Reviewer 3: @GiedriusS we did a few changes in the proposal to make this clear. The TLDR is, we are going to use reflection to discover what services (StoreAPI, TargetsAPI, etc.) a particular endpoint is exposing. The alternative is to expose this information as part of the `Info` service itself. We are leaning towards reflection because in this case if some component adds support for a new service (ex. Store GW adds support for Exemplars) it will be automatically discovered by the Querier without extra work in code.

I think it's worth to explain why we are using reflection in the proposal as well.

> The plan would be for the initial discovery of gRPC services available to happen via gRPC reflection. To get further information, we would be adding an Info method to other services for example, Targets, Exemplars, etc. This would contain similar info or metadata information as we get right now from the Store Info method. We can always add more data to it in the future as needed.

Reviewer: Just to be perfectly clear this time around, I would suggest including a note along the lines of:
```
The following alternatives were also considered, however they were not chosen because of their respective drawbacks.
1. ...
2. ...
```
